src/vscode_ocaml_platform.ml

```diff
@@ -1,263 +1,69 @@
+         the environment. Pick a particular package manager and sandbox by \
+         clicking the button below"
+      ~choices:[ (select_pm_button_text, ()) ]
       ()
   in
-  match selection with
-  | None -> ()
-  | Some () -> select_sandbox instance ()
-
-let register_commands extension =
-  List.iter ~f:(function command, callback ->
-      let callback ~args:_ = callback () in
-      ExtensionContext.subscribe extension
-        ~disposable:(Commands.registerCommand ~command ~callback))
+  Option.iter selection ~f:(Extension_commands.select_sandbox.handler instance)
 
 let activate (extension : ExtensionContext.t) =
+  (* this env var update disables ocaml-lsp's logging to a file
+     because we use vscode [output] pane for logs *)
   Process.Env.set "OCAML_LSP_SERVER_LOG" "-";
-  let instance = Instance.create () in
-  register_commands extension
-    [ (select_sandbox_command_id, select_sandbox instance)
-    ; (restart_command_id, restart_instance instance)
-    ; (open_terminal_command_id, open_terminal instance)
-    ; (open_terminal_select_command_id, select_sandbox_and_open_terminal)
-    ; (switch_impl_intf_command_id, switch_impl_intf instance)
-    ];
-  ExtensionContext.subscribe extension
-    ~disposable:(Instance.disposable instance);
   let open Promise.Syntax in
-  let toolchain =
-    let+ pm = Toolchain.of_settings () in
-    let resources, is_fallback =
-      match pm with
-      | Some toolchain -> (toolchain, false)
-      | None ->
-        let (_ : unit Promise.t) = suggest_to_setup_toolchain instance in
-        (Toolchain.Package_manager.Global, true)
-    in
-    (Toolchain.make_resources resources, is_fallback)
+  let* package_manager =
+    Toolchain.of_settings ()
+    (* TODO: implement [Toolchain.from_settings_or_detect] that would
+       either get the sandbox from the settings or detect in a smart way (not simply Global) *)
+  in
+  let is_fallback = Option.is_empty package_manager in
+  let package_manager =
+    Option.value package_manager ~default:Toolchain.Package_manager.Global
   in
-  let* toolchain, is_fallback = toolchain in
-  Instance.start instance toolchain
-  |> Promise.Result.iter ~error:(fun e ->
-         if not is_fallback then message `Error "%s" e)
+  Extension_instance.make (Toolchain.make package_manager)
+  |> Promise.Result.iter
+       ~ok:(fun instance ->
+         (* register things with vscode, making sure to register their disposables *)
+         let _register_extension_commands : unit =
+           Extension_commands.register_all_commands extension instance
+         in
+         let _register_extension_instance : unit =
+           ExtensionContext.subscribe extension
+             ~disposable:(Extension_instance.disposable instance)
+         in
+         let _register_dune_formatter : unit =
+           Dune_formatter.register instance
+           |> List.iter ~f:(fun disposable ->
+                  ExtensionContext.subscribe extension ~disposable)
```

Comment:
Let's not expose the list of formatters outside of the Dune_formatter module. We can abstract it away as
```ocaml
let dune_formatter = Dune_formatter.register instance in
ExtensionContext.subscribe extension
  ~disposable:(Dune_formatter.disposable dune_formatter)
```